File: GOL.go
//Referred awk code from gist.github/timm
package main

import (
   "fmt"
   "math/rand"
   "os"
   "time"
)

//Function to initialize the matrix with cells
func life(rows int, cols int, threshold float64, generations int) {
   var length int = rows * cols
   var initialState = make([]int, length)
   for i := 0; i < length; i++ {
      if rand.Float64() < threshold {
         initialState[i] = 1
      } else {
         initialState[i] = 0
      }
   }
   live(initialState, rows, generations, length)

}

//Function to generate neighbours based on GOL's 4 rules
func live(currentState []int, rows int, generations int, length int) {

   if generations < 1 {
      os.Exit(0)
   }
   var neighbours int
   fmt.Printf("\033[1;1H")
   fmt.Printf("\033[2J")
   fmt.Printf("Generation %d\n", generations)
   var count int = 0
   for x := 0; x < len(currentState); x++ {
      if currentState[x] == 1 {
         fmt.Print("*")
      } else {
         fmt.Print(" ")
      }
      count += 1
      if (count % rows) == 0 {
         fmt.Println()
      }
   }
   var nextState = make([]int, length)
   for x := 0; x < len(currentState); x++ {
      //-7%2=-1 in golang, so ((x-1)%length+length)%length gives -7%2=1
      neighbours = currentState[((x-1)%length+length)%length] + currentState[((x+1)%length+length)%length] + currentState[((x-rows-1)%length+length)%length] + currentState[((x-rows)%length+length)%length] + currentState[((x-rows+1)%length+length)%length] + currentState[((x+rows-1)%length+length)%length] + currentState[((x+rows)%length+length)%length] + currentState[((x+rows+1)%length+length)%length]
      nextState[x] = currentState[x]
      if currentState[x] == 0 {
         if neighbours == 3 {
            nextState[x] = 1
         } else {
            nextState[x] = 0
         }
      } else {
         if neighbours == 2 || neighbours == 3 {
            nextState[x] = 1
         } else {
            nextState[x] = 0
         }
      }

   }
   generations--
   time.Sleep(time.Second / 2)
   live(nextState, rows, generations, length)
}

//Main function 
func main() {
   life(50, 20, 0.3, 20)
}

File: GOL.rb


def live(now, rows, generations, size)
    if(generations < 1)
        exit 
    end
    print "Generations "+generations.to_s.rjust(3, "0")+"\n"

    for x in 1..(now.length()+1)
        if(now[x] == 1)
            print "*"
        else
            print " "
        end
        if((x%rows) == 0)
            print "\n"
        end
    end
    b = Array.new(size)
    for x in 0..(now.length())
        neighbors = now[((x-1)%size+size)%size]+now[((x+1)%size+size)%size]+
                        now[((x-rows-1)%size+size)%size]+now[((x-rows)%size+size)%size]+
                        now[((x-rows+1)%size+size)%size]+now[((x+rows-1)%size+size)%size]+
                        now[((x+rows)%size+size)%size]+now[((x+rows+1)%size+size)%size]
        b[x] = now[x]
        if(now[x] == 0)
            if(neighbors == 3)
                b[x] = 1
            else
                b[x] = 0
            end
            if(neighbors==2 || neighbors==3)
                b[x] = 1
            else
                b[x] = 0
            end
        end
    end
    generations -= 1
    live(b, rows, generations, size)
end


def life (rows, cols, some, generations)
    size = rows * cols
    now = Array.new(size)
    
    for i in 0..size
        if(rand < some)  
            now[i] = 1
        else
            now[i] = 0
        end
    end
    live(now, rows, generations, size)
end


life(50, 20, 0.619, 20)